src/api/rpc_api/events.rs

```diff
@@ -70,24 +72,65 @@ where
 	}
 }
 
-// FIXME: This should rather be implemented directly on the
-// Subscription return value, rather than the api. Or directly
-// subcribe. Should be looked at in #288
-// https://github.com/scs/substrate-api-client/issues/288#issuecomment-1346221653
+/// Wrapper around a Event `StorageChangeSet` subscription.
+/// Simplifies the event retrieval from the subscription.
+pub struct EventSubscription<Subscription, Hash> {
+	pub subscription: Subscription,
+	_phantom: PhantomData<Hash>,
+}
+
+impl<Subscription, Hash> EventSubscription<Subscription, Hash>
+where
+	Hash: DeserializeOwned,
+	Subscription: HandleSubscription<StorageChangeSet<Hash>>,
+{
+	/// Create a new wrapper around the subscription.
+	pub fn new(subscription: Subscription) -> Self {
+		Self { subscription, _phantom: Default::default() }
+	}
+
+	/// Wait for the next value from the internal subscription.
+	/// Upon encounter, it retrieves and decodes the expected `EventRecord`.
+	pub fn next_event<RuntimeEvent: Decode, Topic: Decode>(
+		&mut self,
+	) -> Option<Result<Vec<EventRecord<RuntimeEvent, Topic>>>> {
+		let change_set = match self.subscription.next()? {
+			Ok(set) => set,
+			Err(e) => return Some(Err(Error::RpcClient(e))),
+		};
+		// Since we subscribed to only the events key, we can simply take the first value of the
+		// changes in the set. Also, we don't care about the key but only the data, so take
+		// the second value in the tuple of two.
+		let storage_data = change_set.changes[0].1.as_ref()?;
+		let events = Decode::decode(&mut storage_data.0.as_slice()).map_err(Error::Codec);
+		Some(events)
+	}
+
+	/// Unsubscribe from the internal subscription.
+	pub fn unsubscribe(self) -> Result<()> {
+		self.subscription.unsubscribe().map_err(|e| e.into())
+	}
+}
+
+impl<Subscription, Hash> From<Subscription> for EventSubscription<Subscription, Hash>
+where
+	Hash: DeserializeOwned,
+	Subscription: HandleSubscription<StorageChangeSet<Hash>>,
+{
+	fn from(subscription: Subscription) -> Self {
+		EventSubscription::new(subscription)
+	}
+}
+
 pub trait SubscribeEvents<Client, Hash>
 where
 	Client: Subscribe,
 	Hash: DeserializeOwned,
 {
-	fn wait_for_event<Ev: StaticEvent>(
```

Comment:
Removed completely. It's overly complex, and with `fetch_events_for ...`, resp. the simplified `subscribe_events` I don't see where this could be useful.